File: src/main/java/datastructures/hashtable/java/CustomHashTable.java
package datastructures.hashtable.java;

import java.util.LinkedList;

public class CustomHashTable<K, V> {
    private LinkedList<Data>[] dataList = new LinkedList[5];

    public boolean put(K key, V value) {
        Integer index = getIndexForKey(key);
        if (dataList[index] == null) {
            LinkedList<Data> bucket = new LinkedList<>();
            Data<K, V> data = new Data<>(key, value);
            bucket.addLast(data);
            dataList[index] = bucket;
            return true;
        }

        for (Data data : dataList[index]) {
            if (data.key.equals(key)) {
                data.value = value;
                return true;
            }
        }
        dataList[index].addLast(new Data<K, V>(key, value));
        return true;
    }

    public boolean remove(K key) {
        Integer index = getIndexForKey(key);
        Data data = get(key);
        if (data != null) {
            dataList[index].remove(get(key));
            return true;
        }
        return false;
    }

    public Object getValue(K key) {
        Integer index = getIndexForKey(key);
        for (Data data : dataList[index]) {
            if (data.key.equals(key)) {
                return data.value;
            }
        }
        return null;
    }


    private Data<K, V> get(K key) {
        Integer index = getIndexForKey(key);
        for (Data data : dataList[index]) {
            if (data.key.equals(key)) {
                return data;
            }
        }
        return null;
    }

    private int getIndexForKey(K key) {
        return key.hashCode() % dataList.length;
    }

    private class Data<K, V> {
        private K key;
        private V value;

        public Data(K key, V value) {
            this.key = key;
            this.value = value;
        }

        public V getValue() {
            return value;
        }
    }
}

File: src/main/java/datastructures/hashtable/kotlin/CustomHashTable.kt
package datastructures.hashtable.kotlin

import java.util.*

class CustomHashTable<K, V> {
    var dataList: Array<LinkedList<Data<K, V>>?> = arrayOfNulls(5)

    data class Data<K, V>(val key: K, var value: V)

    open fun put(key: K, value: V): Boolean {
        val index = getIndexForKey(key)
        if (dataList[index] == null) {
            val bucket = LinkedList<Data<K, V>>()
            bucket.addLast(Data(key, value))
            dataList[index] = bucket
            return true
        }

        for (data in dataList[index]!!) {
            if (data.key == key) {
                data.value = value
                return true;
            }
        }
        dataList[index]?.addLast(Data(key, value))
        return true;
    }

    open fun getValue(key: K): V? {
        val data = get(key)
        if (data != null) return data.value
        return null
    }

    open fun remove(key: K): Boolean {
        val index = getIndexForKey(key)
        val data = get(key)
        if (data != null) {
            dataList[index]?.remove(data)
            return true
        }
        return false
    }

    private fun get(key: K): Data<K, V>? {
        val index = getIndexForKey(key)
        for (data in dataList[index]!!) {
            if (data.key == key) return data
        }
        return null
    }

    private fun getIndexForKey(key: K): Int {
        if (key != null) {
            return key.hashCode() % dataList.size
        }
        return 0;
    }
}